### DISCUSSION TITLE
Is there anyway to specify the order in which code block cells run?

### DISCUSSION THREAD

**Discussion by astrowonk** [quarto-dev/quarto-cli#1837] [Q&A] [answered]
I have a .qmd file (or jupyter notebook) that does a variety of computations but at the end produces a list. I have a TLDR at the top that just gives out the answers that are computed later, but at the moment I have to run all the cells (or render the document), get the list and then cut and paste back in the top.

 Is there any way to essentially run a cell near the top of the document last so it can update automatically/dynamically when rendered? It's not so bad with caching since the second render is fast, but it'd be nice if the opening section could just pull in data that is computed later...
 
 I'm pretty sure this is impossible at the moment...

**Comment by jjallaire** [comment]
You can probably do this but will require some intermediate to advanced hacking. You probably want to write a text file in the final cell that has access to the results and then read it at the top with a Lua filter (either a custom one that you write specifically to read the file and generate the content you want OR using the the `include-files` filter: https://github.com/pandoc/lua-filters/tree/master/include-files

**Comment by rgaiacs** [comment, marked as answer]
Hi @astrowonk what are your constraints?

For what you described, I'm moving my documents to use Observable. Consider the following minimal working example:

``````
---
title: "Reactive Markdown"
---

Example of Observable demonstrating inclusion of result of Python calculation: ${my_output}.

## Long Calculation

```{python}
1 + 1
```

```{python}
my_output = 2
```

```{python}
ojs_define(my_output = my_output)
```
``````

When you render the `.qmd` or `.ipynb` with Quarto, it will process the Python cells and by using the `ojs_define` you will export the value of the variable to Observable. When you open the HTML file, Observable will replace the `${my_output}` with the value exported.

**Reply by astrowonk** [reply]
This looks promising!  But when I ran this exact code block through quarto it just show `${my_output}`, no computation. Where is that inline observable variable documented? I couldn't find it anywhere.

**Reply by astrowonk** [reply]
I need some sort of ojs cell (anywhere that does anything) and then it seems to work. Maybe without that it doesn't know I'm using Observable? or perhaps it's a bug.

**Reply by rgaiacs** [reply]
Can you add at least one `ojs` cell? Without the `ojs` cell, I get

![Screenshot from 2022-08-12 12-22-53](https://user-images.githubusercontent.com/1506457/184284826-a39a633a-7d07-40ca-9273-b1f1807a32c4.png)

With the `ojs` cell, I get

![Screenshot from 2022-08-12 12-22-24](https://user-images.githubusercontent.com/1506457/184284812-fc6fdc21-3438-4bce-b784-cf1d47e311e1.png)

**Reply by rgaiacs** [reply]
I forget to mention where the documentation is: https://quarto.org/docs/computations/ojs.html and https://quarto.org/docs/interactive/ojs/.

**Reply by astrowonk** [reply]
Thanks this is basically working now! I still don't see any reference to these inline observable variables, i.e. `${my_output}`, but this suggestion worked quite well!

I'm getting an odd error in python in the cell that calls (and only calls) `ojs_define`: 

```
Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
```

It only happens in my larger notebook, not in the test one I made from your initial post. I'm just turning off output, warnings, errors and that solved it for now.

The only remaining mystery is why the dummy ojs cell with echo set to off still shows a little widget for code folding (which shows nothing) at the very bottom which you can see at the very bottom in the [live published version.](https://marcos.quarto.pub/wordle_openers/) Thank you!

**Reply by rgaiacs** [reply]
Sorry that my memory is very fog regarding where I saw the inline observable variable. I think I saw people talking about Quarto and Observable on Twitter and in one example they used it.

**Reply by astrowonk** [reply]
Thank you @rgaiacs! If anyone wants to see the final qmd file with this observable trick it is [here](https://github.com/astrowonk/TwitterWordle/blob/main/openers.qmd).

**Reply by cscheid** [reply]
It's a bug that we don't include the `ojs` engine when there are inline OJS expressions but no `ojs` block. Handling this correctly is somewhat subtle and so we are not going to do it for the upcoming 1.1 release, but we'll make sure to fix it. Thanks for catching it!

We're tracking this bug in [this issue](https://github.com/quarto-dev/quarto-cli/issues/1846)
